Library/OcAppleImg4Lib/OcAppleImg4Lib.c

```diff
@@ -0,0 +1,316 @@
+                  NULL,
+                  &DataSize,
+                  &Environment->effectiveProductionStatus
+                  );
+  if (EFI_ERROR (Status)
+   || DataSize != sizeof (Environment->effectiveProductionStatus)) {
+    Environment->effectiveProductionStatus = TRUE;
+  }
+
+  DataSize = sizeof (Environment->effectiveSecurityMode);
+  Status = gRT->GetVariable (
+                  L"EffectiveSecurityMode",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->effectiveSecurityMode
+                  );
+  if (EFI_ERROR (Status)
+   || DataSize != sizeof (Environment->effectiveSecurityMode)) {
+    Environment->effectiveSecurityMode = TRUE;
+  }
+
+  DataSize = sizeof (Environment->internalUseOnlyUnit);
+  Status = gRT->GetVariable (
+                  L"InternalUseOnlyUnit",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->internalUseOnlyUnit
+                  );
+  if (EFI_ERROR (Status)
+   || DataSize != sizeof (Environment->internalUseOnlyUnit)) {
+    Environment->internalUseOnlyUnit = FALSE;
+  }
+
+  // CHANGE: HardwareModel is unused.
+}
+
+/**
+  Verify the signature of ImageBuffer against Type of its IMG4 Manifest.
+
+  @param[in]  This            The pointer to the current protocol instance.
+  @param[in]  ObjType         The IMG4 object type to validate against.
+  @param[in]  ImageBuffer     The buffer to validate.
+  @param[in]  ImageSize       The size, in bytes, of ImageBuffer.
+  @param[in]  SbMode          The requested IMG4 Secure Boot mode.
+  @param[in]  ManifestBuffer  The buffer of the IMG4 Manifest.
+  @param[in]  ManifestSize    The size, in bytes, of ManifestBuffer.
+  @param[out] HashDigest      On output, a pointer to ImageBuffer's digest.
+  @param[out] DigestSize      On output, the size, in bytes, of *HashDigest.
+
+  @retval EFI_SUCCESS             ImageBuffer is correctly signed.
+  @retval EFI_INVALID_PARAMETER   One or more required parameters are NULL.
+  @retval EFI_OUT_OF_RESOURCES    Not enough resources are available.
+  @retval EFI_SECURITY_VIOLATION  ImageBuffer's signature is invalid.
+
+**/
+EFI_STATUS
+EFIAPI
+AppleImg4Verify (
+  IN  APPLE_IMG4_VERIFICATION_PROTOCOL  *This,
+  IN  UINT32                            ObjType,
+  IN  CONST VOID                        *ImageBuffer,
+  IN  UINTN                             ImageSize,
+  IN  UINT8                             SbMode,
+  IN  CONST VOID                        *ManifestBuffer,
+  IN  UINTN                             ManifestSize,
+  OUT UINT8                             **HashDigest OPTIONAL,
+  OUT UINTN                             *DigestSize OPTIONAL
+  )
+{
+  DERReturn          DerResult;
+
+  DERImg4Environment  EnvInfo;
+  DERImg4ManifestInfo ManInfo;
+  UINT8              ImageDigest[OC_MAX_SHA_DIGEST_SIZE];
```

Comment:
Well, there is some indentation issue if you care, and I would still make an ASSERT that ManInfo.imageDigestSize is less or equal to OC_MAX_SHA_DIGEST_SIZE.

Reply:
Would a STATIC_ASSERT that OC_MAX_SHA_DIGEST_SIZE is >= SHA512_DIGEST_SIZE work too or what is the exact concern here?

Reply:
The intention of this check is a desync of OC_MAX_DIGEST_SIZE between the list of digests used in this code. I believe that only runtime assertion of the actual size makes sense here, as it is algorithm independent and future proof.

Reply:
Added at https://github.com/acidanthera/OcSupportPkg/commit/1add52451cea6094760d892d1afddd62e9bc3552